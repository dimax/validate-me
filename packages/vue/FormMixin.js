import { processErrors } from '@validate-me/core/rules';

let errorHandler = f => f;

export function setErrorHandler(handler) {
  errorHandler = handler;
}

export default {
  data() {
    return {
      fields: {},
      invalid: true,
      touched: false,
    };
  },
  provide() {
    return {
      setField: this.setField,
      updateField: this.updateField,
    };
  },
  watch: {
    fields(newFields) {
      for (const state of Object.values(newFields)) {
        if (state) {
          this.invalid = true;

          return;
        }
      }

      this.invalid = false;
    },
  },
  created() {
    this.fieldMethods = {};
  },
  methods: {
    setField(name, methods) {
      if (this.fieldMethods[name]) {
        throw new Error(`Field "${name}" already exists.`);
      }

      this.fieldMethods[name] = methods;
    },
    updateField(name, state) {
      if (this.fields[name] === state) {
        return;
      }
      this.fields = { ...this.fields, [name]: state };
    },
    process(error) {
      return processErrors(this.fieldMethods, errorHandler(error));
    },
    validate() {
      let invalid = true;
      const fields = Object.values(this.fieldMethods);

      if (!this.touched) {
        invalid = fields.reduce(
          (invalid, field) => field.touch() || invalid,
          true,
        );

        this.touched = true;
      }
      if (invalid && this.invalid) {
        return false;
      }
      fields.forEach(field => field.clearWarning());

      return true;
    },
  },
};
