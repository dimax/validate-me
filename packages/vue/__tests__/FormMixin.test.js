/* eslint-disable no-console */
import { mount, createLocalVue } from '@vue/test-utils';
import flushPromises from 'flush-promises';

import { setErrorHandler } from '../FormMixin';
import ValidatemePlugin from '../index';

import Form from './Form';

const localVue = createLocalVue();

localVue.use(ValidatemePlugin);

describe('vue/FormMixin', () => {
  const inputName = 'text';
  const propsData = {
    name: inputName,
    pattern: '^\\D+$',
    rules: [['len', '1', '5']],
  };

  test('Instanciates a form with two fields with the same name', () => {
    // Used to supress the expected error from vue runtime dev
    const err = console.error;

    console.error = () => {};

    expect(() => {
      mount(Form, {
        localVue,
        propsData: { ...propsData, name2: inputName, silent: true },
      });
    }).toThrowErrorMatchingSnapshot();

    console.error = err;
  });

  test('Instanciates a form and validates a valid and invalid input states', async () => {
    const app = mount(Form, { localVue, propsData });

    await flushPromises();

    const input = app.find('input');

    input.element.value = 'a';
    input.trigger('input');

    expect(app.vm.validate()).toMatchSnapshot();

    input.element.value = '1';
    input.trigger('input');

    expect(app.vm.validate()).toMatchSnapshot();
  });

  test('Instanciates a form and process errors from server with(out) custom error handler', async () => {
    const localVue = createLocalVue();

    localVue.use(ValidatemePlugin, {
      errorHandler: f => f,
    });

    const app = mount(Form, { localVue, propsData });

    await flushPromises();

    await app.vm.process({ [inputName]: ['required'] });

    app.setProps({ required: true });
    app.setProps({ required: false });
    expect(app.vm.$children[0].$data).toMatchSnapshot();
    expect(app.vm.validate()).toMatchSnapshot();

    await app.vm.process({ [inputName]: ['pattern', '.+'] });
    expect(app.vm.$children[0].$data).toMatchSnapshot();
    expect(app.vm.validate()).toMatchSnapshot();

    setErrorHandler(err => err.fields);
    await app.vm.process({
      code: 1,
      fields: { [inputName]: ['warningFromServer'] },
    });
    setErrorHandler(f => f);
    expect(app.vm.$children[0].$data).toMatchSnapshot();
    expect(app.vm.validate()).toMatchSnapshot();
  });
});
