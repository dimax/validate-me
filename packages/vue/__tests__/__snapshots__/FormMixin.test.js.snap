// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue/FormMixin Instanciates a form and process errors from server with(out) custom error handler 1`] = `
Object {
  "vField": Object {
    "error": "The field is required.",
    "loading": false,
    "touched": false,
    "value": "",
    "warning": "",
  },
}
`;

exports[`vue/FormMixin Instanciates a form and process errors from server with(out) custom error handler 2`] = `
Array [
  false,
]
`;

exports[`vue/FormMixin Instanciates a form and process errors from server with(out) custom error handler 3`] = `
Object {
  "vField": Object {
    "error": "The field is required.",
    "loading": false,
    "touched": true,
    "value": "",
    "warning": "WARNING: Unknown server rule \\"warningFromServer\\" failed with the value \\"\\"",
  },
}
`;

exports[`vue/FormMixin Instanciates a form and process errors from server with(out) custom error handler 4`] = `
Array [
  false,
]
`;

exports[`vue/FormMixin Instanciates a form and validates a valid and invalid input states 1`] = `
Array [
  true,
  Object {
    "text": "a",
  },
]
`;

exports[`vue/FormMixin Instanciates a form and validates a valid and invalid input states 2`] = `
Array [
  false,
]
`;
